operator/controllers/keptnworkloadinstance/controller.go

```diff
@@ -0,0 +1,179 @@
+//+kubebuilder:rbac:groups=lifecycle.keptn.sh,resources=keptntasks/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;watch
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
+//+kubebuilder:rbac:groups=apps,resources=replicasets,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KeptnWorkloadInstance object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
+func (r *KeptnWorkloadInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log = log.FromContext(ctx)
+	r.Log.Info("Searching for Keptn Workload Instance")
+
+	workloadInstance := &klcv1alpha1.KeptnWorkloadInstance{}
+	err := r.Get(ctx, req.NamespacedName, workloadInstance)
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		r.Log.Error(err, "Workload Instance not found")
+		return reconcile.Result{}, fmt.Errorf("could not fetch KeptnWorkloadInstance: %+v", err)
+	}
+
+	r.Log.Info("Workload Instance found", "instance", workloadInstance)
+
+	if workloadInstance.IsPostDeploymentCompleted() {
+		return reconcile.Result{}, nil
+	}
+
+	r.Log.Info("Post deployment checks not finished")
+
+	isDeployed, err := r.IsWorkloadResourceDeployed(ctx, workloadInstance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if isDeployed {
+		err = r.reconcilePostDeployment(ctx, req, workloadInstance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+	}
+
+	r.Log.Info("deployment not finished")
+
+	if workloadInstance.IsPreDeploymentCompleted() {
+		r.Log.Info("Post deployment checks not finished")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	r.Log.Info("pre-deployment checks not finished")
+
+	err = r.reconcilePreDeployment(ctx, req, workloadInstance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KeptnWorkloadInstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&klcv1alpha1.KeptnWorkloadInstance{}).
+		Complete(r)
+}
+
+func (r *KeptnWorkloadInstanceReconciler) generateSuffix() string {
+	uid := uuid.New().String()
+	return uid[:10]
+}
+
+func (r *KeptnWorkloadInstanceReconciler) IsWorkloadResourceDeployed(ctx context.Context, workloadInstance *klcv1alpha1.KeptnWorkloadInstance) (bool, error) {
```

Comment:
can we rename this to something like `UpdateWorkloadResourceDeploymentStatus()` - otherwise the name is a bit misleading since does not only check for a condition but also updates the workload instance object